_posts/2021-01-12-magic-control.adoc

```diff
@@ -0,0 +1,280 @@
+
+In your pom.xml file, just add the following dependency:
+
+[source, xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-vertx-core</artifactId>
+</dependency>
+----
+
+With this, the _run_ method becomes:
+
+[source, java]
+----
+@Override
+public int run(String... args) {
+    Vertx vertx = CDI.current().select(Vertx.class).get();
+    Router router = Router.router(vertx);
+
+    String message = ConfigProvider.getConfig().getValue("message", String.class);
+
+    router.get("/").handler(rc -> rc.response().end(message + " world!"));
+    router.get("/stop").handler(rc -> {
+        rc.response().end("bye");
+        Quarkus.asyncExit();
+    });
+
+    HttpServer server = vertx.createHttpServer()
+            .requestHandler(router)
+            .listen(8080);
+
+    Quarkus.waitForExit();
+
+    server.close();
+    return 0;
+}
+----
+
+Note how it retrieves the managed Vert.x instance.
+While you can use `@Inject`, you can also retrieve it programmatically.
+The rest of the code does not change.
+We can still start it from the IDE using the _main_ method.
+
+If you don't include the `quarkus-vertx-core` extension (or any extension depending on it), Quarkus won't create the Vert.x instance.
+
+As said above, you can package this application as a native executable:
+
+[source, shell script]
+----
+> mvn package -Dnative
+...
+> ./target/managed-vertx-example-1.0-SNAPSHOT-runner
+----
+
+But, still no hot reload 😿.
+
+== Using the managed HTTP server
+
+Instead of using only the `quarkus-vertx-core` extension, we can choose to delegate the HTTP server to Quarkus.
+You may see that as a loss of control, but actually, we rarely do much around it, and again, you can configure it from the `application.properties` file if needed.
+
+Instead of ``quarkus-vertx-core`, use `quarkus-vertx-http`:
+
+[source, xml]
+----
+<dependencies>
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-vertx-http</artifactId>
+    </dependency>
+</dependencies>
+----
+
+No need to depend on Vert.x Web directly, it's included.
+
+You will still register your routes, but using a managed `Router`:
+
+[source, java]
+----
+@Override
+public int run(String... args) {
+    Router router = CDI.current().select(Router.class).get();
+
+    String message = ConfigProvider.getConfig().getValue("message", String.class);
+
+    router.get("/").handler(rc -> rc.response().end(message));
+    router.get("/bye").handler(rc -> {
+        rc.response().end("bye");
+        Quarkus.asyncExit();
+    });
+
+    Quarkus.waitForExit();
+    return 0;
+}
+----
+
+That approach enables the Quarkus hot reload as it intercepts the HTTP requests.
+You are still in control of everything related to your application logic.
+
+You can start the hot reload using:
+
+[source, shell script]
+----
+> mvn quarkus:dev
+----
+
+== Summary
```

Comment:
Should you have one section where Quarkus is started and not stopped (i.e. not needing the main method?

Reply:
Do you mean having a bean with @Startup?

Reply:
more like the RESTEasy (reactive) like model where the routes are hosted on a class and the server is de facto started. Not sure I'm super clear but it's the full magic one.

BTW this magic business made me thing of It's a kind of magic line by Queen. Do you mind combing your hair and using a Mercury outfit. It would make a good picture for the blog post 👍 

I would do it but you're dark hair like him and I'm not.

Reply:
Ah, I see, yes, adding a section with the "regular" model. Good idea, so it covers the full spectrum.

About your suggestion, unfortunately, I'm way shorter than Mercury (and a lot fatter). I guess you can color your hair. 

Note: This PR review contains a reference to Carla Bruni and a reference to Queen... That's going to be a long year...